Sources/Datadog/Tracing/Utils/URLSessionSwizzler.swift

```diff
@@ -6,102 +6,146 @@
+        let newImpBlock: TypedBlockIMP = { impSelf, impURL, impCompletion -> URLSessionDataTask in
+            let sessionRequest = impSelf.urlRequest(with: impURL)
+            if requiresInterception(sessionRequest) {
+                return impSelf.dataTask(with: sessionRequest, completionHandler: impCompletion)
+            }
+
+            return typedOriginalImp(impSelf, sel, impURL, impCompletion)
         }
         let newImp: IMP = imp_implementationWithBlock(newImpBlock)
         try swizzler.swizzle(selector: sel, in: subjectClass, with: newImp)
     }
 
-    static func swizzleDataTaskWithRequest() throws {
-        typealias TypedIMP = TypedIMPs.DataTaskWithRequest
-        typealias TypedBlockIMP = TypedBlocks.DataTaskWithRequest
-        let sel = Selectors.DataTaskWithRequest
+    static func swizzleDataTaskWithRequestCompletionHandler(
+        requiresInterception: @escaping RequiresInterception,
+        interceptor: @escaping RequestInterceptor,
+        taskObserver: @escaping TaskObserver
+    ) throws {
+        typealias TypedIMP = TypedIMPDataTaskWithURLRequestCompletion
+        typealias TypedBlockIMP = TypedBlockDataTaskWithRequestCompletion
+        let sel = SelectorDataTaskWithRequestCompletion
 
         let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
 
-        let newImpBlock: TypedBlockIMP = { [impSelector = sel] impSelf, impURLRequest -> URLSessionDataTask in
-            return typedOriginalImp(impSelf, impSelector, impURLRequest)
-        }
-        let newImp: IMP = imp_implementationWithBlock(newImpBlock)
-        try swizzler.swizzle(selector: sel, in: subjectClass, with: newImp)
-    }
+        let newImpBlock: TypedBlockIMP = { impSelf, impURLRequest, impCompletion -> URLSessionDataTask in
 
-    static func swizzleDataTaskWithURLCompletionHandler() throws {
-        typealias TypedIMP = TypedIMPs.DataTaskWithURLCompletion
-        typealias TypedBlockIMP = TypedBlocks.DataTaskWithURLCompletion
-        let sel = Selectors.DataTaskWithURLCompletion
+            if requiresInterception(impURLRequest),
+                let interceptionResult = interceptor(impURLRequest) {
 
-        let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
+                weak var blockTask: URLSessionDataTask? = nil
+                let modifiedCompletion: CompletionHandler = { origData, origResponse, origError in
+                    impCompletion(origData, origResponse, origError)
+                    if let completedTask = blockTask {
+                        taskObserver(completedTask, .completed)
+                    }
+                }
+
+                let task = typedOriginalImp(impSelf, sel, interceptionResult.request, modifiedCompletion)
+                blockTask = task
+                taskObserver(task, .created(spanContext: interceptionResult.spanContext))
```

Comment:
`URLSessionSwizzler` shouldn't know about `spanContext`. This leaks the abstraction.

Reply:
yes you are right, that part already took me a while before raising this PR

here is the problem:
1. we need `spanContext` in `task.resume()`
2. therefore, we need to carry `spanContext` from task creation step, after `let task = dataTaskWith...`
3. we create `spanContext` at request creation, in `intercept(originalRequest)` call

in short, we need to carry `spanContext` between 3 non-contagious calling sites.
spilling `DDSpanContext` was the simplest working solution i came up with.
i'll be happy to change if you see a better solution

Reply:
How about replacing closures with `URLSessionDataTaskInterceptor` object, which is attached to `URLSessionTask` through associated-object? Like this:
```swift
class URLSessionDataTaskInterceptor {
    private let configuration: TracedDomains
    private var spanContext: SpanContext?
    private var span: SpanContext?

    /// Called right after the task is created.
    func configure(_ request: inout URLRequest) {
        guard configuration.shouldRequestBeIntercepted(request) else {
            return
        }

        let spanContext = tracer.createSpanContext()
        let headersWriter = DDHTTPHeadersWriter()
        headersWriter.inject(spanContext: spanContext)

        for (headerField, value) in headersWritter.tracePropagationHTTPHeaders {
            request.addValue(value, forHTTPHeaderField: headerField)
        }

        self.spanContext = spanContext
    }

    /// Called when task is resumed (so it starts)
    func notifyTaskResumed() {
        if let spanContext = self.spanContext {
            span = tracer.startSpan(with: spanContext!)
        }
    }

    /// Called when task completes
    func notifyTaskCompleted(data: Data?, response: URLResponse?, error: Error?) {
        if let span = self.span {
            span.setTag("status.code": response.statusCode)
            span.finish()
        }
    }
}
```

This can be easily plugged into existing Swizzler's method with even simplifying the implementation and signature (I did not it 🤚):
```swift
internal class URLSessionSwizzler {
    static func interceptDataTaskWithRequestCompletionHandler(
        using interceptor: URLSessionDataTaskInterceptor
    ) {
        typealias TypedIMP = TypedIMPDataTaskWithURLRequestCompletion
        typealias TypedBlockIMP = TypedBlockDataTaskWithRequestCompletion
        let sel = SelectorDataTaskWithRequestCompletion

        let typedOriginalImp: TypedIMP = try swizzler.currentImplementation(of: sel, in: subjectClass)
        let newImpBlock: TypedBlockIMP = { impSelf, impURLRequest, impCompletion -> URLSessionDataTask in
            var mutableRequest = impURLRequest
            interceptor.configure(&mutableRequest)

            let modifiedCompletion: CompletionHandler = { origData, origResponse, origError in
                impCompletion(origData, origResponse, origError)
                interceptor.notifyTaskCompleted(data: origData, response: origResponse, error: origError)
            }

            let task = typedOriginalImp(impSelf, sel, mutableRequest, modifiedCompletion)

            // associate Interceptor with task, so `interceptor.resume()` can be called appropriately:
            task.interceptor = interceptor

            return task
        }

        let newImp: IMP = imp_implementationWithBlock(newImpBlock)
        try swizzler.swizzle(selector: sel, in: subjectClass, with: newImp)
    }
}
```

This way we can simplify the number of closures and we have a good piece of logic: `URLSessionDataTaskInterceptor` to unit test the span instrumentation logic (without even mocking the `URLSession`).

This also scales up for adding more tracing features (like setting error on `Span` object). For RUM product, we could change `URLSessionDataTaskInterceptor` to protocol and provide two separate implementations (one for tracing, one for RUM), so this is composable.

Reply:
but we never need to mock `URLSession`, do we? 🤔 

there are a few problems with that approach

1. why would `URLSessionTask` need an associated object with `tracedDomains`?
2. we would need to create a new `Interceptor` object per `task`; if we'd like to use a singleton `interceptor` then we would need to implement a `request->task` map or something like that
3. if we group those 3 blocks in the same type, that won't win us much and we would need to mock that type for different scenarios (eg: never intercept, intercept and fail at modifying request, intercept and succeed, resume task with X event, etc.)

if i understand well, you want to proxy `DDSpanContext` with a container type so that `URLSessionSwizzler` won't have any reference to `tracing`, but the code above doesn't address that problem unfortunately

currently, we don't actually need further abstraction. i'm thinking of revisiting this problem when we start implementing swizzling for RUM, when we may need more abstraction for more general use cases.

Reply:
> 1. why would URLSessionTask need an associated object with tracedDomains?

That's why I suggested `TracedDomains` instead of `Set<URL>` so we can pass the `tracedDomains` as a reference type. This will add only the size of the reference pointer to each task - negligible memory impact.

> 2. we would need to create a new Interceptor object per task;

Why this could be a problem? With this approach, the task observation state (`spanContext` + `span`) is centralized within an object, instead of being distributed across closure callbacks. Callbacks force us to leak the abstraction like in the discussed line. The `URLSessionDataTaskInterceptor ` is one way to fix it.

> if we'd like to use a singleton interceptor then we would need to implement a request->task map or something like that

I don't see any reason for `request` → `task` mapping. `Task` becomes the owner of the interceptor (through associated object). No other component component needs to communicate with `URLSessionDataTaskInterceptor`. And even if some does in the future, we can inject a delegate through interceptor `init()`.

> if we group those 3 blocks in the same type, that won't win us much and we would need to mock that type for different scenarios (eg: never intercept, intercept and fail at modifying request, intercept and succeed, resume task with X event, etc.)

Do you mean testing? We can unit test the `URLSessionDataTaskInterceptor` like this:
```swift
func testGivenValidURL_itAddsTracingHeadersToTheRequest() {
   // given
   let url = URL(...)
   let urlRequest: URLRequest = URLRequest(url: url)
   let interceptor = URLSessionDataTaskInterceptor(domains: TracedDomains([url.domain]))

   // when
   interceptor.configure(&urlRequest)

   // then
   XCTAssertNotNil(urlRequest.allHeaderFields["x-trace-id-header"])
   XCTAssertNotNil(urlRequest.allHeaderFields["x-span-id-header"])
}
```
Other scenarios seem easy as well. For the "whole thing" testing, we can leverage integration tests, so no need to even swizzle `URLSession` for unit tests suite (and this is what I meant, when saying "mock `URLSession`", sorry 🙌).

> if i understand well, you want to proxy `DDSpanContext` with a container type so that `URLSessionSwizzler` won't have any reference to tracing, but the code above doesn't address that problem unfortunately

Why doesn't? This gets rid of `spanContext` in `URLSessionSwizzler` at all. And scales up to injecting other types of interceptors if we need some for RUM product. When going with closure-based approach, we'd have to mix `tracingSpanContext` with `rumSpanContext` inside `URLSessionSwizzler` 🙅‍♂️ (both optionals, as features might be disabled / enabled).

> currently, we don't actually need further abstraction. i'm thinking of revisiting this problem when we start implementing swizzling for RUM, when we may need more abstraction for more general use cases.

Disagree. I don't see why we may not want to do scalable piece of architecture now and avoid re-thinking the `URLSession` interception concept when RUM product arrives. I wouldn't highlight this, if we don't see an obvious clash between Tracing and RUM with the current approach. And this clash starts in the discussed line:
```swift
taskObserver(task, .created(spanContext: interceptionResult.spanContext))
```
When adding resources interception for RUM, this will become:
```swift
taskObserver(task, .created(
   tracingSpanContext: tracingInterceptionResult.spanContext,
   rumSpanContext: rumInterceptionResult.spanContext
))
```
Which means we need to re-do `URLSwizzling`. That in turn will be very sensitive and dangerous, as tracing auto-instrumentation will be released and rolled out to production by many apps. Instead, we should leverage this and only once go through the `URLSwizzling` roll-out process.

Reply:
```swift
func swizzle(using interceptor: Interceptor) {
  let newImplementation = {
    interceptor.doSomething()
  }
}
```
if we do that, every call to the swizzled method will use the same `interceptor` object
if we associate this `interceptor` to a `task`, X number of `task`s will share the same `interceptor` object.

we can do this instead:
```swift
func swizzle(using interceptorProvider: (() -> Interceptor)) {
  let newImplementation = {
    let newInterceptor = interceptorProvider()
    newInterceptor.doSomething()
  }
}
```
(or passing `Interceptor.shared` in your proposal and implement a map)
or to keep it simple, we can do this:
```swift
internal struct Payload { // a proxy, container type
  var span: Span?
  var spanContext: SpanContext?
}
...
enum TaskObservationEvent: Equatable {
  case created(payload: Payload)
}
```

or we still can add `RUM` without any change by passing an array `RequestInterceptor`s to swizzled method:
```swift
enum TaskObservationEvent: Equatable {
  case tracingContextAdded(context: DDSpanContext)
  case rumContextAdded(context: RUMContext)
}
...
func swizzle(with interceptors: [RequestInterceptor]) {
  let newImp = {
    let interceptions = interceptors.reduce(originalRequest) { // pass request through interceptor chain }
    let newRequest = interceptions.last.request
    let createdContexts = interceptions.map { $0.context }
    createdContexts.forEach { taskObserver(.contextCreated($0)) }
    ...
  }
}
```

_correction:_ what i mean by revisiting this with RUM is that now we don't have the actual implementation of RUM, we can only educated-guess how it will look, what use-cases we will have. if we make a bad guess now, that can probably lead to a bad abstraction

Reply:
> if we do that, every call to the swizzled method will use the same interceptor object

You're right, I missed this part 👌, but what you propose with:
```swift
func swizzle(using interceptorProvider: (() -> Interceptor))
```
is the easy and correct way to fix it.

> (or passing Interceptor.shared in your proposal and implement a map)

I'd avoid having a map. First, the map would have to be thread safe, so we'd need a `DispatchQueue` to synchronize it, which will open the door for threading issues. Second,  IMO the "one task - one interceptor" abstraction provides much easier mental model to work with.

> or to keep it simple, we can do this: (`struct Payload`)

This is IMO not easier 🤔. This problem requires a state: `spanContext` + `span`. With closures + `Payload` approach, we distribute this state across callbacks, which will be harder to manage.

> what i mean by revisiting this with RUM is that now we don't have the actual implementation of RUM, we can only educated-guess how it will look, what use-cases we will have. if we make a bad guess now, that can probably lead to a bad abstraction

That's why I don't want to go deeply into RUM implementation details 🙌, definitely not 🙅. But what we know even today is that RUM will require separate request interceptor. And what we can quickly grasp from the Android SDK is a visible difference between [`RUMRequestInterceptor`](https://github.com/DataDog/dd-sdk-android/blob/master/dd-sdk-android/src/main/kotlin/com/datadog/android/rum/internal/net/RumRequestInterceptor.kt) and [`TracingRequestInterceptor`](https://github.com/DataDog/dd-sdk-android/blob/master/dd-sdk-android/src/main/kotlin/com/datadog/android/tracing/internal/net/TracingRequestInterceptor.kt). That's why IMO injecting an object handling the interception logic is a way safer strategy than basing the interception on set of tracing-coupled callbacks. Object can be subclassed / abstracted / even duplicated, but callbacks can be only refactored.